fix(RouteRegistry): initialize RouteParams.params

IAPark opened the pull request:
I decided to try to fix [this](https://github.com/angular/angular/issues/3755#issuecomment-133613666). It took me a lot longer than originally expected, but I think I've done it. The root cause seems to have been that Router.generate was changed apparently in this [commit](https://github.com/angular/angular/commit/f66ce096d8d908fcc0a643a14412136e74025c3d) to allow the deep linking from anywhere. It also changed the method from essentially just passing arguments on to the relevant component's RouteRecognizer to generate the arguments from just a list. When this was done it should have always passed at least an empty StringMap for params, but because params is initialized to null this was not the case. 

I had a bit of trouble testing as the testing scripts are not designed to run on windows, but `./node_modules/.bin/protractor protractor-js.conf.js` runs fine with my edits, but throws an exception if I remove just my edits to route_registry.ts.

`./node_modules/.bin/gulp test.js --browsers=${KARMA_BROWSERS:-ChromeCanary}` seems to run fine in either case.

btford commented:
Needs a test.

IAPark commented:
I can do that. I'm going to assume then you'd say I should also remove the check in the constructor of RouteParams for params being null though?

btford commented:
@IAPark – I think it'd be better to check in the constructor and default to `{}` if the param passed in is `null`.

I don't think we want the type error.

IAPark commented:
@btford OK. I almost did it that way anyway and I'll be honest the main reason I threw an exception was because I didn't write tests. 

I'll stick it in router_link_spec, but I'll probably not be able to get to it until tomorrow so if for some reason you want to just do it feel free.

IAPark commented:
Never mind I'm putting it in route_registry_spec.

btford commented:
:+1:

IAPark committed:
fix(RouteRegistry): initialize RouteParams.params

Fix a bug caused by RouteRegistry.generate not initializing RouteParams.params to a StringMap

Closes #3755

IAPark commented:
@btford I think this should do what you want. It definitely fails without my change and passes with it.

I ended up removing the check totally from the constructor of RouteParams. I feel like it's better to let this issue come out quickly if we ever have this happen later rather than silently "fixing it" and it's a bit faster anyway.

IAPark commented:
Oh and by the way I've realized the commit message is slightly messed up/inaccurate, but I think if I change it will run checks again which is a pain as I believe the message can just be changed during merge.

btford commented:
@IAPark – it's fine. It links back to this issue with more context for those interested.

I just pushed it to pre-submit, so it'll be in master soon. Thanks for your help!

btford commented:
Hm, seems like this never landed due to a flake. I'm re-pushing this so it should land this afternoon.

btford commented:
Landed as 3791c4a6827812c983d2cc6af696abaa765cc479.

Thanks @IAPark!
